### PR TITLE
GUI: Skip games during Mass Add

### PR DESCRIPTION
**Pull request by hecmar007**
GUI: Skip games during Mass Add

[x] and [ ] are used as indicators, as well as background color.

### CODE REVIEW

**Comment by bluegr** [comment]
I don't understand... which games are skipped, and why? Can you provide an example?

**Comment by digitall** [comment]
@bluegr : Reading the code, this appears to add the ability to select / deselect games for addition after the Mass Add detection i.e. so you can reject duplicates / false detections etc.

Not sure about the exact details of the implementation UI, but looks not unreasonable. I think this defaults to selecting all detected games i.e. same as current Mass Add behaviour, unless you unselect... which is reasonable.

**Comment by hecmar007** [comment]
@digitall: Yes that is basically it. It's the implementation of the following requested feature: https://bugs.scummvm.org/ticket/14668 . By default all games are selected, as I believe that in most cases users will prefer to add as many games as possible. Then, the user will be able to deselect (or reselect) any duplicates or unwanted games by clicking on the item in the list. [x] next to the game's title and the green highlight background color for the list's item is used to indicate a selected game, while [ ] and the regular background is used to indicate an unselected one. 

I hope this clears things up a bit!

**Comment by chrisis123** [comment]
Since I'm the one who proposed this feature request - yes this is excactly what I proposed. Thank you for picking this up :)

**Comment by tag2015** [comment]
Thanks, this is a useful feature. IMHO keeping all the entries highlighted looks a bit weird:
massadd
personally I'd rather use the `kStateDisabled `text style, like this:
massadd_alt
But it's a matter of preference, the functionality is the same

**Comment by bluegr** [comment]
> Thanks, this is a useful feature. IMHO keeping all the entries highlighted looks a bit weird: massadd personally I'd rather use the `kStateDisabled `text style, like this: massadd_alt But it's a matter of preference, the functionality is the same

I'm also in favor of the disabled entries, without the highlighting - the end result looks more consistent and clean

**Comment by sev-** [comment]
I like the @tag2015 proposal. @hecmar007 could you please adapt your solution?

**Comment by hecmar007** [comment]
No problem, it should be done now! I agree, it does look cleaner this way.

**Comment by i30817** [comment]
My ocd is going crazy because the width of a empty space is not the same as X

File: gui/massadd.cpp [hunk 1]
```diff
@@ -156,11 +159,30 @@ void MassAddDialog::handleCommand(CommandSender *sender, uint32 cmd, uint32 data
...
+void MassAddDialog::updateGameList() {
+	// Update list to correctly display selected / unselected games
+	Common::U32StringArray l;
+	_list->setList(l);
+	_list->clearListSelected();
+
+	for (const auto &game : _games) {
+		Common::U32String displayString = game.isSelected ? Common::String("[x] ") + game.description : Common::String("[ ] ") + game.description;
+		_list->append(displayString);
+		_list->apendListSelected(game.isSelected);
```
Comment (bluegr) [inline gui/massadd.cpp:182]:
```suggestion
		_list->appendListSelected(game.isSelected);
```

File: gui/widgets/list.h [hunk 2]
```diff
@@ -131,6 +133,9 @@ class ListWidget : public EditableWidget {
 	void setEditColor(ThemeEngine::FontColor color) { _editColor = color; }
 	void setFilterMatcher(FilterMatcher matcher, void *arg) { _filterMatcher = matcher; _filterMatcherArg = arg; }
 
+	void apendListSelected(bool selected) { _listSelected.push_back(selected); }
```
Comment (bluegr) [inline gui/widgets/list.h:136]:
```suggestion
	void appendListSelected(bool selected) { _listSelected.push_back(selected); }
```

File: gui/massadd.cpp [hunk 3]
```diff
@@ -156,11 +159,30 @@ void MassAddDialog::handleCommand(CommandSender *sender, uint32 cmd, uint32 data
...
 	} else {
 		Dialog::handleCommand(sender, cmd, data);
 	}
 }
 
+void MassAddDialog::updateGameList() {
+	// Update list to correctly display selected / unselected games
+	Common::U32StringArray l;
+	_list->setList(l);
+	_list->clearListSelected();
```
Comment (bluegr) [inline gui/massadd.cpp:177]:
```suggestion
	_list->clearSelectedFromList();
```

File: gui/widgets/list.h [hunk 4]
```diff
@@ -131,6 +133,9 @@ class ListWidget : public EditableWidget {
...
+	void clearListSelected() { _listSelected.clear(); }
```
Comment (bluegr) [inline gui/widgets/list.h:137]:
```suggestion
	void clearFromSelectedList() { _listSelected.clear(); }
```

**Comment by hecmar007** [comment]
@bluegr squashed your review.

**Comment by tag2015** [comment]
> My ocd is going crazy because the width of a empty space is not the same as X and if offsets the remaining rest of the line from the selected game name 'column'.

The font is not monospace so there's not much that can be done, maybe we could use a bullet for enabled entries and nothing for disabled entries but I'm not sure that's an improvement
Capture

**Comment by i30817** [comment]
I prefer that honestly. Less space wasted for one.

**Comment by hecmar007** [comment]
@tag2015 @i30817 Got it, I agree it looks better. Changes pushed.

**Comment by sev-** [comment]
Does it mean that the list entries are jumping around when you enable/disable them?

**Comment by sev-** [comment]
What you can easily do, is use 'FOUR-PER-EM SPACE' instead of a normal space. We use Unicode there: https://jkorpela.fi/chars/spaces.html

Also, there is point on "saving" horizontal space, as there is plenty of it on the right part of the list.

**Comment by hecmar007** [comment]
@sev- Got it, \u2000 matches x's width. Changes pushed.

**Comment by bluegr** [comment]
> @sev- Got it, \u2000 matches x's width. Changes pushed.

Great! Can we have a screenshot of the end result?

**Comment by hecmar007** [comment]
> > @sev- Got it, \u2000 matches x's width. Changes pushed.
> 
> Great! Can we have a screenshot of the end result?

Here you go:
image

**Comment by sev-** [comment]
Thank you!

**Comment by i30817** [comment]
One thing about this bothers me. A lot of duplicates are from deficiencies of the scanner falsely recognizing a directory as the whole game when the truth is the parent directory is the whole game, USECASE dir in Ultima 8 and crusader games is one of the examples. You also can't depend on a consistent order of evaluation (depth vs breadth first) because I've seen both.

This would be more useful if there was a tooltip with the path on hover is what I'm saying.

**Comment by sev-** [comment]
This is a completely irrelevant issue. Detection tables for ultima should be fixed and include paths, so it stays on the top.

**Comment by i30817** [comment]
If only it was just Ultima.

**Comment by sev-** [comment]
The people never report these things, that's why

**Comment by sluicebox** [comment]
This breaks colors in other list dialogs. The disabled-item color is now used for all items.

BrowserDialog:
image

ChooserDialog:
image

**Comment by tag2015** [comment]
> AGS has more than a few

Use a daily build and report these to the bugtracker, please

**Comment by tag2015** [comment]
I can confirm the issue sluicebox reported. Please check. but i think It should be enough to do
```
// Display selected/unselected games in mass detection as enabled/disabled items.
_state = ThemeEngine::kStateEnabled;
if (pos < (signed int)_listSelected.size() && !_listSelected[pos])
    _state = ThemeEngine::kStateDisabled;
```

**Pull request merged**